### ISSUE TITLE
Plugin assets conflicting with DebugKit

### ISSUE THREAD

**Issue by shahways** [cakephp/cakephp#14568]
This is a (multiple allowed):

* [x] bug

* CakePHP Version: 4.0.7
* Platform and Target: 
DebugKit 4.1.2
macOS Catalina 10.15.4
MAMP with Apache 2.2.34,  PHP 7.4.2 and MySQL 5.7.26

### What you did
- Baked a theme-only plugin and added a 'templates' and 'webroot' folder. 
- Created a default.php layout and placed some image, css and js assets in the webroot folder of the plugin. 
- Activated the theme in the AppController in the beforeRender callback. 
- Created a view in the plugin in templates/Pages/test.php

### What happened
- The page keeps pending on the plugin assets (js/img/css). After a minute all the assets return a 500 error. 
- The error.log shows "Error: [PDOException] SQLSTATE[HY000]: General error: 5 database is locked", the stack trace shows 2 lines related to DebugKit (vendor/cakephp/debug_kit/src/ToolbarService.php:256 and vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php:66)
- Loading a plugin asset directly through the URL works fine and fast
- Deleting the tmp/debug_kit.sqlite file had no effect
- When debug set to false, everything works perfectly

### What you expected to happen
When debugging is enabled, plugin assets should be loading just as fast without locking the DebugKit sqlite db. 

### Additional information
In #757 (https://github.com/cakephp/debug_kit/pull/757) at the bottom @allanmcarvalho apparently describes the same situation, but I could not find any related issues here. His screenshot of his browser inspector looks exactly like mine.

**Comment by markstory** [comment]
> When debugging is enabled, plugin assets should be loading just as fast without locking the DebugKit sqlite db.

Plugin assets are going through the application middleware stack and thus get recorded by DebugKit. If sqlite locking is proving to be a problem, you could try moving DebugKit to a different RDBMS like Postgres, MySQL or SQLServer.

**Comment by shahways** [comment]
Thanks for the insight Mark. Still completely new to 4.x and we are currently looking into plugins to rebuild our current 2.x codebase. Apparently DebugKit has become quite more advanced and thus also more complex. Just bizarre that requesting a Plugin asset directly through URL works as it should. Also assets in the main application work fine. The problem is only manifested when a plugin layout / view is in use.

**Comment by othercorey** [comment]
Is it possible to post the stack trace?

**Comment by shahways** [comment]
Yes of course, here you go: 

2020-05-08 23:32:17 Error: [PDOException] SQLSTATE[HY000]: General error: 5 database is locked in /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Statement/StatementDecorator.php on line 178
Stack Trace:
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Statement/StatementDecorator.php:178
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Statement/BufferedStatement.php:150
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Statement/SqliteStatement.php:41
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Connection.php:342
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Core/Retry/CommandRetry.php:69
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Connection.php:345
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Query.php:245
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:2036
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:1928
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:1829
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:1529
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Database/Connection.php:692
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:1530
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/ORM/Table.php:1830
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/debug_kit/src/ToolbarService.php:256
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php:66
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Http/Runner.php:73
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Http/Runner.php:58
- /Applications/MAMP/htdocs/domain.com/vendor/cakephp/cakephp/src/Http/Server.php:90
- /Applications/MAMP/htdocs/domain.com/webroot/index.php:40

**Comment by markstory** [comment]
> Apparently DebugKit has become quite more advanced and thus also more complex. Just bizarre that requesting a Plugin asset directly through URL works as it should. Also assets in the main application work fine. The problem is only manifested when a plugin starts calling assets through the HtmlHelper.

Loading an asset individually and as part of a page change the concurrency at which your application handles requests. Sqlite databases generally don't support concurrent access. If you are not frequently adding/changing your plugin assets you could use the `cake plugin assets symlink` command to side step CakePHP handling the plugin assets.

**Comment by othercorey** [comment]
I think we could catch this kind of error and ignore it. There are probably other scenarios were there are concurrent requests.

**Comment by shahways** [comment]
@markstory Yes, we were planning to symlink assets, but only on production, based on the performance recommendations in the cookbook.

If my understanding is correct, this issue only affects plugin assets. I tested by creating a new view in /templates/Pages/test.php (so outside of the plugin). I then copied the assets (image + css + js) from the plugin webroot to the main app webroot, and referenced them through the HtmlHelper in the default.php layout. The URL /pages/test loads fine, no sqlite blocking issues with DebugKit. 

So DebugKit seems to handle application assets fine, but chokes when it comes to plugin assets. Is this possible?

**Comment by ADmad** [comment]
> I then copied the assets (image + css + js) from the plugin webroot to the main app webroot, and referenced them through the HtmlHelper in the default.php layout.

Assets in your app's webroot would be served directly by the webserver, CakePHP wouldn't be involved at all.

**Comment by shahways** [comment]
> > I then copied the assets (image + css + js) from the plugin webroot to the main app webroot, and referenced them through the HtmlHelper in the default.php layout.
> 
> Assets in your app's webroot would be served directly by the webserver, CakePHP wouldn't be involved at all.

Ah yes @ADmad, the .htaccess ensures that of course, my bad. Ok, curious to see if a structural solution can be provided in the 4.0.8 release!

**Comment by allanmcarvalho** [comment]
I'm not sure if it's for the same reason, but I also had this problem for plugin routes ... when I try to access a url inside the plugin and the debug is on, it causes the same error.
